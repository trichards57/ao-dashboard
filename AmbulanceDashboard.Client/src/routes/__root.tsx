// -----------------------------------------------------------------------
// <copyright file="__root.tsx" company="Tony Richards">
// Copyright (c) Tony Richards. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for
// full license information.
// </copyright>
// -----------------------------------------------------------------------

import { Outlet, createRootRoute } from "@tanstack/react-router";
import { Suspense, lazy } from "react";

import Layout from "../layout";

const TanStackRouterDevtools =
  process.env.NODE_ENV === "production"
    ? () => null // Render nothing in production
    : lazy(() =>
        import("@tanstack/router-devtools").then((res) => ({
          default: res.TanStackRouterDevtools,
        })),
      );

// eslint-disable-next-line import/prefer-default-export
export const Route = createRootRoute({
  component: () => (
    <Layout>
      <Outlet />
      <Suspense>
        <TanStackRouterDevtools />
      </Suspense>
    </Layout>
  ),
});
