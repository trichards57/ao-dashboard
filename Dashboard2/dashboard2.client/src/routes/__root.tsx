import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
import {
  createRootRouteWithContext,
  Outlet,
  useRouteContext,
} from "@tanstack/react-router";
import React from "react";
import Navbar from "../components/navbar";
import { UserInfo, userMeOptions } from "../queries/user-queries";
import { QueryClient, useSuspenseQuery } from "@tanstack/react-query";

const TanStackRouterDevtools =
  process.env.NODE_ENV === "production"
    ? () => null
    : React.lazy(() =>
        import("@tanstack/router-devtools").then((res) => ({
          default: res.TanStackRouterDevtools,
        })),
      );

const Root = () => {
  const meQuery = useSuspenseQuery(userMeOptions);
  const me = meQuery.data;
  const { canEditRoles, canEditVehicles, canViewUsers, canViewVor } =
    useRouteContext({ strict: false });

  return (
    <>
      <Navbar
        canEditRoles={canEditRoles}
        canEditVehicles={canEditVehicles}
        canViewUsers={canViewUsers}
        canViewVor={canViewVor}
        loggedIn={me !== null}
        name={me?.realName}
      />
      <main className="container is-fluid">
        <Outlet />
      </main>
      <TanStackRouterDevtools />
      <ReactQueryDevtools />
    </>
  );
};

export interface RootContext extends UserInfo {
  queryClient: QueryClient;
  loggedIn: boolean;
  canEditRoles: boolean;
  canEditVehicles: boolean;
  canViewUsers: boolean;
  canViewVor: boolean;
}

export const Route = createRootRouteWithContext<RootContext>()({
  loader: (o) => {
    return o.context.queryClient.ensureQueryData(userMeOptions);
  },
  component: Root,
});
