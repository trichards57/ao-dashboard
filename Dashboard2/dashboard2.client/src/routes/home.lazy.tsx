import { createLazyFileRoute } from '@tanstack/react-router'

export const Route = createLazyFileRoute('/home')({
  component: () => <div>Hello /home!</div>
})