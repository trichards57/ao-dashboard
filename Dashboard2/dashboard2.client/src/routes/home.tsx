import { createFileRoute } from '@tanstack/react-router'

export const Route = createFileRoute('/home')({
  component: () => <div>Hello /home!</div>
})