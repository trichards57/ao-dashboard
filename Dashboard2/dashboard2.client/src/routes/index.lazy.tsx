import {
  Navigate,
  createLazyFileRoute,
  useRouteContext,
} from "@tanstack/react-router";

const Index = () => {
  const context = useRouteContext({ from: "/" });

  console.log(context);

  if (context.loggedIn) {
    return <Navigate to="/home" />;
  }

  return (
    <section className="hero">
      <div className="hero-body">
        <p className="title">Welcome</p>
        <p className="subtitle">
          Welcome to the AO Ambulance Dashboard. You will need to log in to use
          this application.
        </p>
      </div>
      <div className="hero-foot is-flex is-justify-content-center">
        <a className="button is-primary" href="/Identity/Account/Login">
          Log In
        </a>
      </div>
    </section>
  );
};

export const Route = createLazyFileRoute("/")({
  component: Index,
});
