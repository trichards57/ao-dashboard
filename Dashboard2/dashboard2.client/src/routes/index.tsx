import { createFileRoute, redirect } from "@tanstack/react-router";
import { useTitle } from "../components/useTitle";

export function Index() {
  useTitle();

  return (
    <section className="hero">
      <div className="hero-body">
        <p className="title">Welcome</p>
        <p className="subtitle">
          Welcome to the AO Ambulance Dashboard. You will need to log in to use
          this application.
        </p>
      </div>
      <div className="hero-foot is-flex is-justify-content-center">
        <a className="button is-primary" href="/Identity/Account/Login">
          Log In
        </a>
      </div>
    </section>
  );
}

export const Route = createFileRoute("/")({
  beforeLoad: ({ context }) => {
    if (context.loggedIn) {
      throw redirect({
        to: "/home",
      });
    }
  },
  component: Index,
});
