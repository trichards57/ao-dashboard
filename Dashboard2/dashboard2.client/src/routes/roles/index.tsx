import { createFileRoute, Link } from "@tanstack/react-router";
import { preloadRoles, useRoles } from "../../queries/role-queries";
import { useTitle } from "../../components/useTitle";
import {
  redirectIfLoggedOut,
  redirectIfNoPermission,
} from "../../support/check-logged-in";

export function Roles() {
  const { data } = useRoles();
  useTitle("User Roles");

  const roles = [...data].sort((a, b) => a.name.localeCompare(b.name));

  return (
    <>
      <h1 className="title">User Roles</h1>

      <table className="table is-striped is-fullwidth">
        <thead>
          <tr>
            <th>Role</th>
            <th>Vehicle Configuration</th>
            <th>VOR Data</th>
            <th>User Role</th>
            <th>Role Permissions</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {roles.map((r) => (
            <tr key={r.id}>
              <td>{r.name}</td>
              <td>{r.vehicleConfiguration}</td>
              <td>{r.vorData}</td>
              <td>{r.permissions}</td>
              <td>{r.name == "Administrator" ? "Write" : "Deny"}</td>
              <td className="edit">
                {r.name != "Administrator" && (
                  <Link to="/roles/edit/$id" params={{ id: r.id }}>
                    Edit
                  </Link>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
}

export const Route = createFileRoute("/roles/")({
  component: Roles,
  loader: ({ context }) => preloadRoles(context.queryClient),
  beforeLoad: ({ context }) => {
    redirectIfLoggedOut(context);
    redirectIfNoPermission(context.canEditRoles);
  },
});
