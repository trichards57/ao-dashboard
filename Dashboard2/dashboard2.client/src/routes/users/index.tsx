import { createFileRoute, Link, redirect } from "@tanstack/react-router";
import { preloadUsers, useUsers } from "../../queries/user-queries";
import { useTitle } from "../../components/useTitle";

function Users({ userId, isAdmin }: { userId: string; isAdmin: boolean }) {
  const { data } = useUsers();

  useTitle("User Settings");

  const sorted = data
    .toSorted((a, b) => a.name.localeCompare(b.name))
    .filter((r) => r.role !== "Administrator" || isAdmin);

  return (
    <>
      <h1 className="title">User Settings</h1>

      <table className="table is-striped is-fullwidth">
        <thead>
          <tr>
            <th>User</th>
            <th>Role</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {sorted.map((r) => (
            <tr key={r.id}>
              <td>{r.name}</td>
              <td>{r.role ?? "None"}</td>
              <td className="edit">
                {r.id.toUpperCase() !== userId.toUpperCase() && (
                  <Link to={`/users/edit/${r.id}`}>Edit</Link>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
}

export const Route = createFileRoute("/users/")({
  component: function Component() {
    return (
      <Users
        userId={Route.useRouteContext().userId}
        isAdmin={Route.useRouteContext().isAdmin}
      />
    );
  },
  loader: ({ context }) => preloadUsers(context.queryClient),
  beforeLoad: ({ context }) => {
    if (!context.loggedIn) {
      throw redirect({
        to: "/",
      });
    }
    if (!context.canViewUsers) {
      throw redirect({
        to: "/home",
        search: { region: "All", district: "All", hub: "All" },
      });
    }
  },
});
