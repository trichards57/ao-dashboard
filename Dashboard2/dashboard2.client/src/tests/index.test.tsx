import { act, render } from "@testing-library/react";
import ReactDOM from "react-dom/client";
import { describe, expect, it } from "vitest";

import { Index } from "../routes";

describe("Index Component", () => {
  it("renders without error", () => {
    const root = ReactDOM.createRoot(document.createElement("div"));
    expect(() => {
      act(() => {
        root.render(<Index />);
        root.unmount();
      });
    }).not.toThrow();
  });

  it("renders correctly", () => {
    const { asFragment } = render(<Index />);
    expect(asFragment()).toMatchSnapshot();
  });
});
