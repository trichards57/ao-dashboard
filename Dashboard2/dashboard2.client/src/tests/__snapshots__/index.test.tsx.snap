// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Index Component > renders correctly 1`] = `
<DocumentFragment>
  <section
    class="hero"
  >
    <div
      class="hero-body"
    >
      <p
        class="title"
      >
        Welcome
      </p>
      <p
        class="subtitle"
      >
        Welcome to the AO Ambulance Dashboard. You will need to log in to use this application.
      </p>
    </div>
    <div
      class="hero-foot is-flex is-justify-content-center"
    >
      <a
        class="button is-primary"
        href="/Identity/Account/Login"
      >
        Log In
      </a>
    </div>
  </section>
</DocumentFragment>
`;
